src/needleman_wunsch.jl

```diff
@@ -58,25 +58,107 @@ end
-    setupmatrix = zeros(Int, length(s1)+1, length(s2)+1)
+    """sets up a matrix with dimension (rows, columns) where # of rows is length of s1 + 1 and # of columns is length of s2 + 1. Dictionary at position 1,1 will have 0 as a key.
+    Dictionaries in the first row: dictionary at pos 1,2 will have the gap score as a key; for every dictionary to its right, the gap score will be incrementing by itself for each dictionary
+    moving to the right. Dictionaries in the first column: dictionary at pos 2,1 will have the gap score as a key; for every dictionary below it, the gap score will be inccrementing by itself
+    for each dictionary moving down. All header dictionaries will have the value nothing. All other dictionaries are empty and of type Any. 
+    """
+    #set up a matrix of all zero's
+    #setupmatrix = zeros(Int, length(s1)+1, length(s2)+1)
+    setupmatrix = Array{Any}(undef, length(s1)+1, length(s2)+1)
+    #make the horizontal and vertical "headers" with the appropriate gap scores
     for j in 1:length(s2)+1
-        setupmatrix[1, j] = (j-1)*gap
+        setupmatrix[1, j] = Dict((j-1)*gap=>nothing)  
     end
     for i in 1:length(s1)+1
-        setupmatrix[i,1] = (i-1)*gap
+        setupmatrix[i,1] = Dict((i-1)*gap=>nothing)
+    end
+    #make the rest of the cells empty dictionaries
+    for i in 2:length(s1)+1
+        for j in 2:length(s2)+1
+            setupmatrix[i, j] = Dict()
+        end
     end
     return setupmatrix
 end
 
+#baseMat = nwsetupmatrix("AGCT", "AGGCC")
+#display(baseMat)
+
+
 function nwscorematrix(seq1, seq2; match=1, mismatch=-1, gap=-1)
-    scoremat = nwsetupmatrix(seq1, seq2; gap=gap)
+    """Revised version of nwscorematrix(). All cells except the headers are dictionaries, where keys are scores and values are lists of string(s) give
+    the direction (e.g. "Above", "Left", "Diagonal") from which the score came from."""
+    scoremat = nwsetupmatrix(seq1, seq2;  gap=gap)
     for i in 2:size(scoremat, 1) # iterate through row indices
         for j in 2:size(scoremat, 2) # iterate through column indices
-            above = nwscore(seq1[i-1], nothing; match, mismatch, gap) + scoremat[i-1,j]
-            left = nwscore(nothing, seq2[j-1]; match, mismatch, gap) + scoremat[i,j-1]
-            diagonal = nwscore(seq1[i-1], seq2[j-1]; match, mismatch, gap) + scoremat[i-1,j-1]
-            scoremat[i,j] = max(above, left, diagonal)
+            
+            # if i == 2 && j ==2
+            #     above = nwscore(seq1[i-1], nothing; match=match, mismatch=mismatch, gap=gap) + scoremat[i-1,j]
+            #     left = nwscore(nothing, seq2[j-1]; match=match, mismatch=mismatch, gap=gap) + scoremat[i,j-1]
+            #     diagonal = nwscore(seq1[i-1], seq2[j-1]; match=match, mismatch=mismatch, gap=gap) + scoremat[i-1,j-1]
+            # else
+            above = nwscore(seq1[i-1], nothing; match=match, mismatch=mismatch, gap=gap) + collect(keys(scoremat[i-1,j]))[1]
```

Comment:
This is pretty inefficient, since you're allocating an array the size of `keys(scoremat...)` on each loop, then discarding it. You could instead do `first(keys(scoremat...))` rather then `collect`ing and indexing